### PR TITLE
Subkeys 2

### PR DESCRIPTION
**Pull request by maurolacy**
Improved test cases after previous PR comments.

Added the two remaining test cases.

### CODE REVIEW

**Comment by maurolacy** [comment]
Still need to test `execute()`. Guess I'll need an integration test for it... will be working on it asap.

File: contracts/cw1-subkeys/src/contract.rs [hunk 1]
```diff
@@ -251,14 +252,142 @@ mod tests {
...
+        setup_test_case(
+            &mut deps,
+            &env,
+            &admins,
+            &initial_spenders,
+            &initial_allowances,
+            &initial_expirations,
+        );
+
+        // Check allowances work for accounts with balances
```
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:285]: Nice tests

File: contracts/cw1-subkeys/src/contract.rs [hunk 2]
```diff
@@ -251,14 +252,142 @@ mod tests {
...
+            config,
+            ConfigResponse {
+                admins: initial_admins.clone(),
+                mutable: true,
+            }
+        );
+
+        // Add a third (new) admin
+        let msg = HandleMsg::UpdateAdmins {
+            admins: [&config.admins[..], &[admin3.clone()]].concat(),
```
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:332]: Nice use of the spread and concat operators. But somehow, I would prefer to see this more explicit.

```rust
let new_admins = vec![owner.clone(), admin2.clone(), admin3.clone()]
let msg = HandleMsg::UpdateAdmins { admins: new_admins.clone() };
// ...
assert_eq!(config, ConfigResponse { admins: new_admins, mutable: true });
```

This is the exact same functionality you have written but easier to read. And key in test cases is legibility (performance is not an issue)

**Comment by ethanfrey** [comment]
Going through this code again, not the PR, I notice there is no test for execute, that the allowances work as they are supposed to.

**Comment by maurolacy** [comment]
Yes. I'll write the `execute()` tests tomorrow.

File: contracts/cw1-subkeys/src/contract.rs [hunk 3]
```diff
@@ -646,4 +646,72 @@ mod tests {
...
+        assert_eq!(res.log, vec![log("action", "execute"), log("owner", spender1)]);
+
+        // And then cannot (not enough funds anymore)
+        let res = handle(&mut deps, env, handle_msg.clone());
+        match res.unwrap_err() {
+            StdError::Underflow { .. } => {},
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // Owner / admins can do anything (at the contract level)
```
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:713]: Good test case.

Can you add another where the spender (with allowance) tries to send a message besides `BankMsg::Send`. Anything else to show it fails.

Just one more test and this is :+1:

Comment (maurolacy) [reply contracts/cw1-subkeys/src/contract.rs:713]: Will do.

**Pull request merged**